rllib/algorithms/algorithm.py

```diff
@@ -669,6 +669,9 @@ def setup(self, config: AlgorithmConfig) -> None:
                     "either a class path or a sub-class of ray.rllib."
                     "offline.offline_evaluator::OfflineEvaluator"
                 )
+            # TODO (Rohan138): Refactor this and remove deprecated methods
+            # Need to add back method_type in case Algorithm is restored from checkpoint
+            method_config["type"] = method_type
```

Comment:
Can you elaborate on what the root of the problem was? When checkpointing, what gets checkpointed? and why is "type" not part of that dict already? We should not pop() and then add back "type" in, that's not a good programming pattern :)

Reply:
Unfortunately currently the design of the ope_dict looks like {"estimator_name": {"type": estimator_class, "estimator_arg0": arg0, ...}} where the args are passed in directly to the dict. This means we need to pop "type" when instatiating the estimator, since "type" defines the estimator class itself, and is not an argument to the constructor, then add it back to the config dict.

IIUC we use a similar pattern for the Replay Buffer and Algorithm "type" keys elsewhere, we should eventually put in a PR that refactors all usages of this antipattern at once.

Reply:
I see. So do we really pop("type") and insert it back in for replay_buffers? can you link the code?

Reply:
We do something similar for ReplayBuffer using the from_config utility on this [line](https://github.com/ray-project/ray/blob/master/rllib/algorithms/algorithm.py#L2641) of algorithm.py; [from_config](https://github.com/ray-project/ray/blob/master/rllib/utils/from_config.py) deepcopies the entire config dict and pops the type off instead of popping it and adding it back.

Reply:
ok. I see. sounds good. Let's go with this.